pkg/kn/commands/root.go

```diff
@@ -16,19 +16,44 @@ package commands
 
 import (
 	"fmt"
+	"io"
 	"os"
 	"path/filepath"
 
+	serving "github.com/knative/serving/pkg/client/clientset/versioned/typed/serving/v1alpha1"
 	homedir "github.com/mitchellh/go-homedir"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 var cfgFile string
 var kubeCfgFile string
 
+// Parameters for creating commands. Useful for inserting mocks for testing.
+type KnParams struct {
+	Output         io.Writer
+	ServingFactory func() (serving.ServingV1alpha1Interface, error)
+}
+
+func (c *KnParams) Initialize() {
+	if c.ServingFactory == nil {
+		c.ServingFactory = GetConfig
+	}
+}
+
 // rootCmd represents the base command when called without any subcommands
-func NewKnCommand() *cobra.Command {
+func NewKnCommand(params ...KnParams) *cobra.Command {
+	var p *KnParams
+	if len(params) == 0 {
+		p = &KnParams{}
+	} else if len(params) == 1 {
+		p = &params[0]
+	} else {
+		panic("Too many params objects to NewKnCommand")
+	}
+	p.Initialize()
```

Comment:
callee of NewKnCommand should probably call Initialize() instead of this constructor knowing about it. as a nice side effect tests wont keep on calling it every time as well.

Reply:
Disagree. The goal is to minimize the interface, so when you call NewKnCommand you're ready to go.